### PR TITLE
Adjust segment feature to be compatible with the SuluArticleBundle

### PR DESCRIPTION
**Pull request by nnatter**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Related issues/PRs | #5277
| License | MIT

#### What's in this PR?

This PR adjusts the segments feature implemented in #5277 to make it usable for articles. Articles are not assigned to a single webspace, therefore this PR needs to adjust the `segment_select` content type to allow for selecting a segment for each webspace.

The PR adjust the following classes to use the new data-format for storing assigned segments:
* `SegmentSelect` field type (+ newly created `segment_select` content type)
* `NavigationQueryBuilder`
* `ContentRouteProvider`
* `QueryBuilder` used for page smart content data provider
* `BlockContentType`

Furthermore, the PR adjusts the `RouteProvider` of the `RouteBundle` to switch segments if the object assigned to the route sets a segment. This is needed to switch segments when accessing an article that is assigned to a specific segment.

#### ToDo

- [ ] Write migrations for segments field

### CODE REVIEW

**Comment by nnatter** [comment]
This PR does not adjust the `article` smart-content data-provider (which is not part of the repository). To fully implement the segment feature for articles, the [`ArticleDataProvider`](https://github.com/sulu/SuluArticleBundle/blob/master/Content/ArticleDataProvider.php) of the `SuluArticleBundle` needs to be adjusted.

File: src/Sulu/Component/SmartContent/ContentType.php [hunk 1]
```diff
@@ -275,7 +275,7 @@ public function getContentData(PropertyInterface $property)
         $limit = (\array_key_exists('limitResult', $filters) && $configuration->hasLimit()) ?
             $filters['limitResult'] : null;
         $options = [
-            'webspaceKey' => $property->getStructure()->getWebspaceKey(),
+            'webspaceKey' => $this->requestAnalyzer->getWebspace()->getKey(),
```
Comment (nnatter) [inline src/Sulu/Component/SmartContent/ContentType.php:278]: the article structure is not assigned to any webspace. therefore `$property->getStructure()->getWebspaceKey()` will always be null if we use a smart-content property inside of an article template

File: src/Sulu/Bundle/PageBundle/Resources/js/containers/Form/fields/SegmentSelect.js [hunk 2]
```diff
@@ -1,43 +1,90 @@
...
+        onChange(webspace.key, value);
+    };
+
+    render() {
+        const {disabled, value, webspace} = this.props;
+
+        return (
+            <div className={segmentSelectStyles.webspaceSection}>
+                <label className={fieldStyles.label}>
+                    {translate('sulu_admin.segments')} ({webspace.name})
```
[the remark this answers is not in the document]
Comment (danrot) [reply src/Sulu/Bundle/PageBundle/Resources/js/containers/Form/fields/SegmentSelect.js:30]: Just talked to @chirimoya, we are going to keep it as it is, just changing the order of the text.

File: src/Sulu/Bundle/PageBundle/Resources/config/forms/page_excerpt.xml [hunk 3]
```diff
@@ -37,14 +37,10 @@
             </meta>
         </property>
         <property
-            name="ext/excerpt/segment"
+            name="ext/excerpt/segments"
             type="segment_select"
-            visibleCondition="__webspace ? __webspace.segments|length &gt; 0 : false"
-        >
-            <meta>
-                <title>sulu_admin.segments</title>
```
[the remark this answers is not in the document]
Comment (danrot) [reply src/Sulu/Bundle/PageBundle/Resources/config/forms/page_excerpt.xml:45]: See comment at https://github.com/sulu/sulu/pull/5425#discussion_r471346731, we are handling that differently now.

File: src/Sulu/Bundle/PageBundle/Tests/Unit/Content/Types/SegmentSelectTest.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+ */
+
+namespace Sulu\Bundle\PageBundle\Tests\Unit\Content\Types;
+
+use PHPCR\NodeInterface;
+use PHPUnit\Framework\TestCase;
+use Sulu\Bundle\PageBundle\Content\Types\SegmentSelect;
+use Sulu\Component\Content\Compat\PropertyInterface;
+
+class SegmentSelectTest extends TestCase
```
Comment (nnatter) [inline src/Sulu/Bundle/PageBundle/Tests/Unit/Content/Types/SegmentSelectTest.php:19]: i think this misses a test for the `remove()` method

**Review by nnatter — COMMENTED** [review]
looks good otherwise 🙂

**Comment by danrot** [comment]
@nnatter done

**Comment by danrot** [comment]
Some tests are failing, I am investigating the reason for this.

**Comment by danrot** [comment]
Seems like the `#` does not work in a query with Jackrabbit. Going to replace that character with `-`.

**Comment by danrot** [comment]
@nnatter Fixed tests as well :slightly_smiling_face:

**Pull request merged**